custom_metrics/privacy.js

```diff
@@ -15,10 +39,145 @@ return JSON.stringify({
-    (typeof window.__cmp == 'function') |
-    (typeof window.__tcfapi == 'function'),
+    // description of `__tcfapi`: https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20CMP%20API%20v2.md#how-does-the-cmp-provide-the-api
+    let tcData;
+    if (typeof window.__tcfapi == 'function') {
+      // based on https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20CMP%20API%20v2.md#gettcdata
+      // Test site: rtl.de
+      __tcfapi('getTCData', 2, (result, success) => {
+        if(success) {
+          tcData = result;
+          tcData.compliant_setup = true;
+        } else {
+          // special case for consentmanager ('CMP settings are used that are not compliant with the IAB TCF')
+          // see warning at the top of https://help.consentmanager.net/books/cmp/page/changes-to-the-iab-cmp-framework-js-api
+          // cf. https://help.consentmanager.net/books/cmp/page/javascript-api
+          // Test site: ?
+          __tcfapi('noncompliant_getTCData', 2, (result, success) => {
+            if(success) {
+              tcData = result;
+              tcData.compliant_setup = false;
+            } else {
+              tcData = {"error": "Failed to retrieve TCData"};
+            }
+          });
+        }
+      });
+    } else {
+      tcData = null;
+    }
+    return tcData;
+  })(),
+  iab_usp: (() => {
+    // IAB US Privacy User Signal Mechanism “USP API” is integrated on the page.
+    // https://github.com/InteractiveAdvertisingBureau/USPrivacy
+    // Test site: nfl.com
+    let uspData;
+    if (typeof __uspapi == 'function') {
+      // based on https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20CMP%20API%20v2.md#gettcdata
+      __uspapi('getUSPData', 1, (result, success) => {
+        if(success) {
+          uspData = result;
+        } else {
+          uspData = {"error": "Failed to retrieve USPData"};
+        }
+      });
+    } else {
+      uspData = null;
+    }
+    return uspData;
+  })(),
+
+  // Ads Transparency Spotlight Data Disclosure schema
+  // Check `meta` tag cf. https://github.com/Ads-Transparency-Spotlight/documentation/blob/main/implement.md
+  // Only for top frame; can't access child frames (same-origin policy)
+  // Test site: ?
+  ads_transparency_spotlight: (() => {
+    let ats_meta_tag = document.querySelector('meta[name="AdsMetadata"]');
+    if (ats_meta_tag) {
+      return ats_meta_tag.content; // JSON message with metadata
+    } else {
+      return null;
+    }
+  })(),
+
+  // FLoC origin trial
+  // Check `meta` tag cf. https://developer.chrome.com/blog/floc/ (complement with relevant header later on)
+  // Test site: https://floc-ot-meta.glitch.me/
+  floc_origin_trial: (() => {
```

Comment:
@VictorLeP How can we know that a specific trial tag is related to FLoC?
All of the meta tags look similar to me, with just a token being different for each trial/origin.

See here: https://developer.chrome.com/origintrials/#/trials/my

Reply:
You're right, the presence of the token just indicates that the website takes part in an origin trial, not which one specifically. It's possible to extract this from the token:
```js
let tag = "AnQAa+uudTrMiqy0/XRpOyGD9+1P1giQjrAWmPiXNjjzPGt7Tg1pjYGKR4+Do80oKP3ltWfkHhxF3Aq317+8gw0AAABheyJvcmlnaW4iOiJodHRwczovL2Zsb2Mtb3QtbWV0YS5nbGl0Y2gubWU6NDQzIiwiZmVhdHVyZSI6IkludGVyZXN0Q29ob3J0QVBJIiwiZXhwaXJ5IjoxNjI2MjIwNzk5fQ==";
window.atob(content);
> "\u0002t\u0000kë®u:Ì¬´ýti;!÷íOÖ\b°\u0016ø68ó<k{N\riG£Í((ýåµgä\u001e\u001cEÜ\n·×¿¼\r\u0000\u0000\u0000a{\"origin\":\"https://floc-ot-meta.glitch.me:443\",\"feature\":\"InterestCohortAPI\",\"expiry\":1626220799}"
```
The `feature` in the included JSON indicates the type of trial (`InterestCohortAPI` = FLoC trial).

I'm wondering whether the trial token could be useful for any other chapter (other trials might be revelant for them), or whether we should focus on FLoC trial tokens only. Then, we also have the choice to process the token upfront (in a custom metric, for both `meta` tags and headers) or to process them after the fact using BigQuery, to extract the `feature`. What are your thoughts?

Reply:
I chatted with a colleague today, and some of the trials might also be useful for at least the Security chapter. I'm thinking of creating a separate custom metrics file (`origin_trial.js`) just for this data point, such that all chapters can use it (without it belonging to the 'privacy' namespace).

Reply:
Nice, following your example I found an example decoding the token:
- https://googlechrome.github.io/OriginTrials/check-token.html

Will remove it from FLoC and add a generic origin-trials metric.

Reply:
Sorry, I haven't reloaded the page and haven't seen your comment about origin_trial.js when writing mine one.
Did you start working on it?

Reply:
Not yet, let me know if you prefer to work on it yourself, else I can do it :)

Reply:
Yeah, I've started it yesterday. Will push today.